internal/metrics/metrics.go

```diff
@@ -0,0 +1,185 @@
+package metrics
+
+import (
+	"os"
+	"time"
+
+	gitopsSyncCommon "github.com/argoproj/gitops-engine/pkg/sync/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/numaproj-labs/numaplane/pkg/apis/numaplane/v1alpha1"
+)
+
+type MetricsServer struct {
+	syncCounter               *prometheus.CounterVec
+	gitSyncUpdateErrorCounter *prometheus.CounterVec
+	reconcileHistogram        *prometheus.HistogramVec
+	kubeRequestCounter        *prometheus.CounterVec
+	kubectlExecCounter        *prometheus.CounterVec
+	gitRequestCounter         *prometheus.CounterVec
+	gitRequestFailedCounter   *prometheus.CounterVec
+	gitRequestLatency         *prometheus.HistogramVec
+	monitoredKubeAPI          *prometheus.GaugeVec
+	kubeCachedResource        *prometheus.GaugeVec
+	kubeCacheFailed           *prometheus.CounterVec
+	hostname                  string
+}
+
+var (
+	descAppDefaultLabels = []string{"name", "namespace"}
+	syncCounter          = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "numaplane_app_sync_total",
+			Help: "Number of application syncs.",
+		},
+		append(descAppDefaultLabels, "dest_namespace", "dest_cluster", "phase"),
+	)
+
+	gitSyncUpdateErrorCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "numaplane_git_sync_update_error",
+			Help: "Error on updating GitSync status.",
+		},
+		descAppDefaultLabels,
+	)
+
+	reconcileHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "numaplane_app_reconcile_duration_second",
```

Comment:
nit: `numaplane_gitsync_reconcile_duration_second` and `GitSync reconciliation performance`